### DISCUSSION TITLE
en/blog/sidecar-injection-iptables-and-traffic-routing/

### DISCUSSION THREAD

**Discussion by giscus** [rootsongjc/rootsongjc.github.io#296] [Announcements]
# en/blog/sidecar-injection-iptables-and-traffic-routing/

This blog describes the sidecar pattern, transparent traffic hijacking and routing.

https://jimmysong.io/en/blog/sidecar-injection-iptables-and-traffic-routing/

**Comment by liuqi-sun** [comment]
inbound handler 中inbound|9080|| 集群type为ORIGINAL_DST，envoy的处理逻辑是不是会通过getsockopt系统调用获取原始目的地地址和port，然后通过127.0.0.6作为socket的源地址，发送请求到通过getsockopt系统调用获取的原始目的地地址，即pod 地址

**Comment by zhuliang1122** [comment]
宋老师，对于应用出站流量，我有一点疑问没搞懂。打个比方product微服务既要访问数据库，又要访问review微服务，iptables的ISTIO_OUTPUT规则链，怎么区分这个两种类型的流量的呢？希望宋老师能够解惑，谢谢！

**Reply by liuqi-sun** [reply]
区分不了，都被劫持到15001端口，envoy会透传非servicemesh 服务的流量

**Reply by zhuliang1122** [reply]
透传非servicemesh流量的意思是：除了调用其他微服务的流量，其他流量都会透传吗？那如果目的流量是tls或者双向tls,envoy还能做到透传吗？我觉得是不是做不到。

**Reply by liuqi-sun** [reply]
其他都透传，比如访问baidu.com。如果你在业务容器内访问baidu.com，envoy就会拦截，一般是根据目的端口找对应的假listener，每一个假listener中又关联路由，路由一般都是端口号，每个路由中都会有一个allow_any的virtual_host，这个就是用来透传。如下图：
![image](https://user-images.githubusercontent.com/48784730/192216590-46cd93b5-daf1-4e0d-8e6d-71c87134a432.png)
业务容器内访问curl http://www.baidu.com
envoy access log：
![image](https://user-images.githubusercontent.com/48784730/192216825-4eb43875-a1d1-4254-9ef3-ca570337a463.png)
业务容器内访问curl https://www.baidu.com
envoy access log：
![image](https://user-images.githubusercontent.com/48784730/192217881-c9b75537-e0c2-4301-9bb2-fca959016ca2.png)

https看日志，我理解应该走的tcp透传，因为我的配置里没有443端口的路由。

virtualOutbound listener use_original_dst为true，而且有一个tcp fliter，集群为PassthroughCluster
![image](https://user-images.githubusercontent.com/48784730/192217429-28500ba2-b613-4943-9d63-db20bf7752ed.png)
![image](https://user-images.githubusercontent.com/48784730/192217797-d5c26122-72dc-465f-90fd-1e32587259d1.png)

**Reply by zhuliang1122** [reply]
如果sidecar容器故障故障，在sidecar容器停止到启动期间，因为应用容器访问非mesh的流量都要通过sidecar容器透传，那应用容器访问baidu.com、数据库不能正常访问。请问下我的理解对吗？

**Reply by zhuliang1122** [reply]
> 区分不了，都被劫持到15001端口，envoy会透传非servicemesh 服务的流量

envoy怎么判断流量是非servicemesh服务的流量，这个能解惑一下吗？我猜可能是要和coreDNS配合

**Reply by liuqi-sun** [reply]
> > 区分不了，都被劫持到15001端口，envoy会透传非servicemesh 服务的流量
> 
> envoy怎么判断流量是非servicemesh服务的流量，这个能解惑一下吗？我猜可能是要和coreDNS配合

istio会生成servicemesh服务的配置(也就是envoy的配置)，还有透传的配置，envoy根据请求匹配不到配置就透传

**Reply by liuqi-sun** [reply]
> 如果sidecar容器故障故障，在sidecar容器停止到启动期间，因为应用容器访问非mesh的流量都要通过sidecar容器透传，那应用容器访问baidu.com、数据库不能正常访问。请问下我的理解对吗？

是的，sidecar故障了，当然就不能提供服务了，但是iptables还在生效，所以请求会失败

**Comment by hanxiantao** [comment]
Istio透明流量劫持示意图中，Outbound流量路由15的虚线是不是画错了，个人感觉应该是ISTIO_OUTPUT->POSTROUTING->外部

**Reply by gabriel9895** [reply]
我觉得也是，规则4，7 如果请求是由 Envoy 代理发出的，则返回 OUTPUT 继续调用 POSTROUTING 规则，最终直接访问目的地。

**Comment by huanglu20124** [comment]
有个疑问，为啥不只拦截目标端口是9080的流量，而是等流量到envoy里再做透传？这样应该能减少不必要的转发吧。
